Прохожу спринт 10

EntityVadim opened the pull request:
Ревью, прошу тебя внести комментарии по коду, может, где-то, что-то стоит поменять, улучшить и т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутые подскажешь.
Буду тебе очень сильно в этом благодарен!

EntityVadim committed:
Создал в проекте файл Constants.swift, в нём объявил перечисление enum Constants, в созданном перечислении объявил статические константы accessKey, secretKey и redirectURI и сохранил в них Access key, Secret Key, Redirect URI своего приложения; добавил константу accessScope — список доступов; добавил константу defaultBaseURL и сохраните в неё базовый адрес API Unsplash

EntityVadim committed:
Добавил в Assets изображение auth_screen_logo, в  main добавил View Controller, добавил Image View и Button на это View, добавил необходимые констрейнты для Image View и Button; установил скругление для кнопки через User Defined Runtime Attributes, создал группу Auth, и затем в этой группе создал файл AuthViewController.swift и в нём объявил класс AuthViewController, в main задал для созданного ранее View Controller'а класс AuthViewController и встроил AuthViewController в Navigation Controller

EntityVadim committed:
Создал в папке Auth новый файл WebViewViewController.swift, добавил в main новый View Controller, сделал сегвей от кнопки Войти``AuthViewController к WebViewViewController, добавил WebView и задал констрейнты, добавил @IBOutlet в WebViewViewController и импорт фреймворка WebKit, доработал кнопку «Назад»

EntityVadim committed:
Чтобы загрузить веб-контент, нужно сформировать запрос URLRequest — для этого потребуется URL, сформировал его с помощью URLComponents; Для успешной авторизации поможет навигационный делегат WKNavigationDelegate, сделал WebViewViewController навигационным делегатом для webView; Обработал получение кода, это не входит в ответственность WebViewViewController, поэтому я делегировал обработку

EntityVadim committed:
Организовал код, дал более понятные названия методам

EntityVadim committed:
Обновил цвета в main, так как поменял их названия в Assets

EntityVadim committed:
Добавил в main UIProgressView, выставил отступы и добавил IBOutlet в WebViewViewController

EntityVadim committed:
реализовал KVO в WebViewViewController

EntityVadim committed:
Создал функцию, которая создаст URLRequest для получения авторизационного токена; Создал файл и класс OAuth2Service; Добавил функцию, которая создаст URLRequest; Добавил функцию fetchOAuthToken с входным параметром code; Добавил в её реализацию создание URLRequest с помощью подготовленного метода и выполните запрос с помощью URLSession.shared.dataTask.

EntityVadim committed:
Изменил параметры видимости методов

EntityVadim committed:
Закончил реализацию fetchOAuthToken и дописал часть логики работы с авторизационным токеном.
Что нужно сделать:
1. Добавить структуру для декодинга JSON-ответа от Unsplash — OAuthTokenResponseBody.
2. Добавить создание URLSessionTask с получением статус-кода, тела ответа и ошибок с помощью расширения URLSession (возможна реализация с помощью NetworkClient).
Блок completion должен быть вызван на главном потоке (DispatchQueue.main). Это нужно, чтобы избежать ошибок при попытке обновить UI не из главного потока.
3. Добавить в fetchOAuthToken второй аргумент completion.
-Замыкание completion должно возвращать Swift.Result<String, Error>, где success соответствует только ситуации, когда в ответе на запрос был получен положительный HTTP-статус-код (значение в диапазоне 200 − 299), а в теле ответа получен и декодирован авторизационный токен.
4. В случае выполнения запроса нужно сохранить Bearer Token в User Defaults.
-Для этого сначала создайте новый класс OAuth2TokenStorage с вычислимым свойством token, для которого переопределены сеттер (функция set) и геттер (функция get).
-Переопределённые сеттер и геттер должны сохранять значение и читать его из User Defaults.
5. Добавить логирование ошибок, полученных в запросе токена. Должны быть распечатаны в консоль:
сетевые ошибки;
-ошибки, которые вернул сервис Unsplash;
-ошибка, которую может выкинуть декодер при получении OAuthTokenResponseBody.
6. В классе AuthViewController, там, где вызывается метод делегата webViewViewController(_ vc: WebViewViewController, didAuthenticateWithCode code: String), вызвать fetchAuthToken.

EntityVadim committed:
Splash: самостоятельная задача
Ваше самостоятельное задание на SplashViewController будет таким:
- Создайте стартовый контроллер, который проверяет, был ли ранее авторизован пользователь.
- Если есть сохранённые данные об авторизации, загрузите необходимую информацию о пользователе и направьте его на флоу галереи. Если таких данных нет — отправьте на флоу авторизации.
- При первом запуске приложения пользователь должен увидеть SplashViewController, а затем — форму авторизации.
- При попадании на флоу авторизации пользователь остаётся на экране, пока не завершит сценарий. Пользователю доступна навигация «Назад» с WebView и повторная попытка авторизоваться с AuthViewController.

НЕ получается пункт:
- При повторном запуске (после ранее успешной авторизации) пользователь сразу после SplashViewController попадёт в галерею.

EntityVadim committed:
Улучшил обработку ошибок

EntityVadim committed:
Struct поменял на enum в Constants; добавил две приватные константы в AuthViewController; Организовал код

EntityVadim commented:
> Ревью, прошу тебя внести комментарии по код, может, где-то, то-то стоит поменять, улучшить і т.д. или ссылки на полезные статьи, как я могу изменить свой код, может какие-то лайфхаки крутыекажеш подсь.Буду тебе очень сильно в этом благодарен! 
Дорогой ревью, я сдаюсь, я не смог реализовать этот пункт:
- Если POST-запрос завершается успешно — экран авторизации скрывается, а в приложении открывается экран с лентой.
Или же я просто не правильно понял задание! Помоги и подскажи мне, что я не так сделал!

EntityVadim committed:
Внес мелкие коррективы (Перенес url ссылки в отдельный файл, исправил ошибку "Value 'token' was defined but never used; consider replacing with boolean test")

EntityVadim committed:
Обновил обработку ошибок

EntityVadim committed:
Я дважды использовал метод fetchOAuthToken и это было ошибкой, я ее исправил

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    private let showWebViewSegueIdentifier = "ShowWebView"
+    private let oauth2Service = OAuth2Service.shared
+    private let oauth2TokenStorage = OAuth2TokenStorage()
```
Эта константа и следующая ниже не используются в классе, давай избавимся от них

So1ovyov commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,115 @@
+    
+    // MARK: - ObserveValue
+    
+    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
```
Предлагаю такие конструкции располагать компактнее, чтобы визуально в ширину не увеличивать код:
```
    override func observeValue(
        forKeyPath keyPath: String?,
        of object: Any?,
        change: [NSKeyValueChangeKey : Any]?,
        context: UnsafeMutableRawPointer?
    ) {
```

So1ovyov commented on ImageFeed/Constants.swift:
```diff
@@ -0,0 +1,29 @@
+
+// MARK: - Constants
+
+enum Constants {
```
Отличный подход, разделить на несколько `enum`-ов логически.

So1ovyov commented on ImageFeed/NetworkServices/OAuth2Service.swift:
```diff
@@ -0,0 +1,78 @@
+    
+    // MARK: - Token Request
+    
+    private func createTokenRequest(withCode code: String) -> URLRequest {
```
Я бы рекомендовал тогда и приватный метод вынести в расширение:
```
// MARK: - OAuth2Service

final class OAuth2Service {
    
    static let shared = OAuth2Service()
    
    init() {}
}

// MARK: - Private methods

private extension OAuth2Service {

    func createTokenRequest(withCode code: String) -> URLRequest {
        var urlComponents = URLComponents(string: Constants.urlComponentsURLString)!
        urlComponents.queryItems = [
            URLQueryItem(name: "client_id", value: Constants.accessKey),
            URLQueryItem(name: "client_secret", value: Constants.secretKey),
            URLQueryItem(name: "redirect_uri", value: Constants.redirectURI),
            URLQueryItem(name: "code", value: code),
            URLQueryItem(name: "grant_type", value: "authorization_code")
        ]
        guard let tokenURL = urlComponents.url else {
            fatalError("Invalid URL components.")
        }
        var tokenRequest = URLRequest(url: tokenURL)
        tokenRequest.httpMethod = "POST"
        return tokenRequest
    }
}

// MARK: - Public methods

extension OAuth2Service {
```

So1ovyov commented on ImageFeed/NetworkServices/OAuth2Service.swift:
```diff
@@ -0,0 +1,78 @@
+                completion(.failure(NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Не удалось получить HTTP-ответ."])))
+                return
+            }
+            switch httpResponse.statusCode {
```
Круто получилось, обработал все кейсы

So1ovyov commented on ImageFeed/NetworkServices/OAuth2Service.swift:
```diff
@@ -0,0 +1,78 @@
+    
+    static let shared = OAuth2Service()
+    
+    init() {}
```
Забыл `private`

So1ovyov commented on ImageFeed/NetworkServices/OAuth2TokenStorage.swift:
```diff
@@ -0,0 +1,23 @@
+
+// MARK: - OAuth2TokenStorage
+
+final class OAuth2TokenStorage {
```
Если это синглтон то забыл `private init() {}`
https://www.waldo.com/blog/swift-singleton

So1ovyov commented on ImageFeed/NetworkServices/OAuthTokenResponseBody.swift:
```diff
@@ -0,0 +1,31 @@
+        self.createdAt = createdAt
+    }
+    
+    enum CodingKeys: String, CodingKey {
```
Для того, чтобы не работать с snake_case ключами, не описывать их в CodingKeys, ты можешь также 
настраивать JSONDecoder под тот вид данных, который приходит от сервера. 
Рекомендую изучить этот способ самостоятельно https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,88 @@
+    
+    // MARK: - Private Properties
+    
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
```
Наименование констант, переменных и методов должно быть в стиле CamelCase и начинаться с нижнего регистра.

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,88 @@
+    // MARK: - Private Properties
+    
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
+    private let oAuthTokenResponseBody = OAuthTokenResponseBody()
```
Это не используется в классе, давай удалим.

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,88 @@
+    private let ShowAuthenticationScreenSegueIdentifier = "ShowAuthenticationScreen"
+    private let oAuthTokenResponseBody = OAuthTokenResponseBody()
+    private let oAuth2TokenStorage = OAuth2TokenStorage()
+    private let oAuth2Service = OAuth2Service()
```
Это также здесь лишнее, предлагаю удалить

So1ovyov commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -0,0 +1,88 @@
+    // MARK: - Private Methods
+    
+    private func switchToTabBarController() {
+        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene else { fatalError("Could not find window scene") }
```
Обычно для конструкций guard часть с else переносят:
```
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene else {
            fatalError("Could not find window scene")
        }
```

So1ovyov commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,79 @@
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == showWebViewSegueIdentifier {
+            guard
```
Обычно такой формат используют, но твой тоже валидный
```
            guard let webViewViewController = segue.destination as? WebViewViewController else {
                fatalError("Failed to prepare for \(showWebViewSegueIdentifier)")
            }
```

EntityVadim committed:
Исправил замечания ревью, а именно:
- Приватный метод createTokenRequest сделал расширением
- Все синглтоны в проекты, сделал с приватным инициализатором
- Для того, чтобы не работать с snake_case ключами, не описывать их в CodingKeys, я настроил JSONDecoder под тот вид данных, который приходит от сервера.
- Переименовал некоторые константы в стиле camelCase
- Удалил некоторые константы, которые не использовались в моем коде (oAuthTokenResponseBody, oAuth2TokenStorage, oAuth2Service)
- Сделал компоновку кода
- В классе OAuth2Service все блоки completion теперь вызваны на главном потоке (DispatchQueue.main) (пункт 10)

So1ovyov commented on ImageFeed/NetworkServices/OAuth2Service.swift:
```diff
@@ -38,39 +39,73 @@ final class OAuth2Service {
         let task = URLSession.shared.dataTask(with: tokenRequest) { data, response, error in
             if let error = error {
-                completion(.failure(error))
+                DispatchQueue.main.async {
```
Чтобы во всем методе не дублировать `DispatchQueue.main.async {}` было бы круто реализовать расширение, которое `completion`-ы сразу оборачивает в асинхронный вызов на главном потоке:
```
enum NetworkError: Error {
    case httpStatusCode(Int)
    case urlRequestError(Error)
    case urlSessionError
}

extension URLSession {
    func data(
        for request: URLRequest,
        completion: @escaping (Result<Data, Error>) -> Void
    ) -> URLSessionTask {
        let fulfillCompletionOnTheMainThread: (Result<Data, Error>) -> Void = { result in
            DispatchQueue.main.async {
                completion(result)
            }
        }

        let task = dataTask(with: request) { data, response, error in
            if let data = data, let response = response, let statusCode = (response as? HTTPURLResponse)?.statusCode {
                if 200 ..< 300 ~= statusCode {
                    fulfillCompletionOnTheMainThread(.success(data))
                } else {
                    print("HTTP error: \(statusCode)")
                    fulfillCompletionOnTheMainThread(.failure(NetworkError.httpStatusCode(statusCode)))
                }
            } else if let error = error {
                print("URL request error: \(error)")
                fulfillCompletionOnTheMainThread(.failure(NetworkError.urlRequestError(error)))
            } else {
                print("URL session error")
                fulfillCompletionOnTheMainThread(.failure(NetworkError.urlSessionError))
            }
        }

        task.resume()
        return task
    }
}
```

Ну или просто применить в своем методе:
```
        let fulfillCompletionOnTheMainThread: (Result<Data, Error>) -> Void = { result in
            DispatchQueue.main.async {
                completion(result)
            }
        }
```
